rfcs/NNNN-preact-signals.md

```diff
@@ -0,0 +1,184 @@
+- A directive that applies a single signal to a binding
+- A customized `html` template tag that automatically applies the directive to signal-values objects.
+
+### SignalWatcher Mixin
+
+Conceptually, we want to run the reactive update lifecycle in an effect so that the signal library observes access to signals and trigger a new update.
+
+We can do this with an override of `performUpdate()` that wraps ReactiveElement's implementation in an effect:
+
+```ts
+    private _disposeEffect?: () => void;
+
+    override performUpdate() {
+      // ReactiveElement.performUpdate() also does this check, so we want to
+      // also bail early so we don't erroneously appear to not depend on any
+      // signals.
+      if (!this.isUpdatePending) {
+        return;
+      }
+      // If we have a previous effect, dispose it
+      this._disposeEffect?.();
+
+      // We create a new effect to capture all signal access within the
+      // performUpdate phase (update, render, updated, etc) of the element.
+      this._disposeEffect = effect(() => {
+        // When Signals change we need to re-render, but we need to get past
+        // the isUpdatePending in performUpdate(), so we set it to true.
+        this.isUpdatePending = true;
+        // We call super.performUpdate() so that we don't create a new effect
+        // only as the result of the effect running.
+        super.performUpdate();
+      });
+    }
+```
+
+An issue with this approach is that we bypass the reactive update lifecycle when signals change. We would like to integrate updates from signals and reactive properties into one batch. To do this we need separate watch and update code paths. Preact Signals don't have such an API, but we should be able to simulate it:
+
+```ts
+    override performUpdate() {
+      if (!this.isUpdatePending) {
+        return;
+      }
+      this._disposeEffect?.();
+      let updateFromLit = true;
+      this._disposeEffect = effect(() => {
+        if (updateFromLit) {
+          super.performUpdate();
+        } else {
+          // This branch is an effect run from Preact signals.
+          // This will cause another call into performUpdate, which will
+          // then create a new effect watching that update pass.
+          this.requestUpdate();
+        }
+        updateFromLit = false;
+      });
+    }
+```
+
+### watch() directive
+
+The `watch()` async directive accepts a signal and renders its value synchronously to the containing binding. When the signal changes, the binding value is updated directly.
+
+Usage:
+```ts
+html`<p>${watch(messageSignal)}</p>`
+```
+
+```ts
+class WatchSignal extends AsyncDirective {
+  render(signal: Signal<unknown>) {
+    let updateFromLit = true;
+    effect(() => {
+      // Access the signal unconditionally to watch it
+      const value = signal.value;
+      if (!updateFromLit) {
+        this.setValue(value);
+      }
+      updateFromLit = false;
+    });
+    return signal.value;
```

Comment:
Hi, @justinfagnani. I am new in this context of trying to contribute, so please, be patient with me. 🙏 

Maybe we need to subscribe to signal changes since this method tracks state and effect automatically.

```ts
class WatchSignal extends AsyncDirective {
  render(signal: Signal<unknown>) {
    let updateFromLit = true;
    signal.subscribe((value) => {
      if(!updateFromLit) {
        this.setValue(value)
      }
      updateFromLit = false;
    })
  }
}
```

It is based on `subscribe` implementation

https://github.com/preactjs/signals/blob/3c23713a5de6e083d549ffb005759914f5dd5304/packages/core/src/index.ts#L274-L286

What do you think ?

Reply:
Thanks @redfox-mx 

I'm really not sure if we should use `subscribe()` instead of `effect()`, since there's no docs for subscribe. Mayeb @developit could advise?

I'll try it anyway in the implementation.